cupy/core/carray.pxi

```diff
@@ -122,9 +122,60 @@ cpdef str _get_header_source():
     return _header_source
 
 
+cdef str _convert_cuComplex_to_Thrust(str source):
+    cdef str macro, line
+    cdef list source_lines = [line for line in source.split('\n')]
+
+    # First, we comment out the line that includes cuComplex.h
+    for i, line in enumerate(source_lines):
+        if '#include' in line and 'cuComplex.h' in line:
```

Comment:
Well, this approach sounds fragile and may surprise users. Do we really need to support `cuComplex.h`? How about providing a bridge interface in the separate header file (in that case users need to replace `cuComplex.h` to sth like `cupy/cuComplex_bridge.h`)?

Reply:
Thanks for raising your concern, @kmaehashi.

> Do we really need to support `cuComplex.h`?

Short answer: Yes. Please do. 

Long answer: #2111 is just a long list of the cases I have encountered that request this. Please do not surprise our users. Many scientific projects need complex numbers, but some existing CUDA projects either do not want to depend on Thrust, or it's too large to do a clean overhaul to transit to Thrust's complex types. So, when people port code to CuPy, they are often frustrated. I've helped many people resolve this, and I think this PR solves the issue once and for all. (So people would stop bothering their local CuPy expert, i.e., me...)

If you are really, really against such a support, can you please at least merge #1866? It's hanging there for quite a while, and it's really frustrating. In either case, we need to make it clear to the users. No surprises.

> How about providing a bridge interface in the separate header file (in that case users need to replace `cuComplex.h` to sth like `cupy/cuComplex_bridge.h`)?

I do not think this is less fragile than the current Pythonic approach. First, this requires code change (although it's only one line), which is against the design principle I had in mind when pushing for `RawModule`. (Users should just import the source code and let CuPy take care of the rest to make it work.)

Second, in the end the macros in this PR go into `cupy/cuComplex_bridge.h`, but the net effect, after the C preprocessing, is identical. So I don't see why we wanna bother users to make a change?

If you are still concerned and not fully convinced, may I propose to mark the new `enable_cuComplex` flag as experimental and subject to change? Let me reiterate that the changes in this PR are not enabled by default, so normal use cases would just continue to work (as you can see from the Jenkins). It would not hurt to consider this PR.

Thanks again, @kmaehashi.

Reply:
@kmaehashi just check in. Any comment? Thanks.

Reply:
OK, now I understand your motivation to use cuComplex code directly on CuPy RawModule/RawKernel.

* Regarding the option name, how about `translate_cucomplex` instead of `enable_cuComplex`? (1) It is difficult for users to imagine that the code is automatically rewritten when `enable` is set. (2) pep8 recommends lowercase for variable names.
* Could you make the option a keyword-only argument so that we can easily consider reorder/renaming later?
* Please use more stricter regex that matches the whole line to detect a header file. You need to consider cases like: `# include <cuComplex.h>` or `// #include <cuComplex.h>`
* Could you extract `macro` contents as a separate file (`cupy/cuComplex_bridge.h`)?
  Then replace `#include <cuComplex.h>` with `#include <cupy/cuComplex_bridge.h>  // translate_cucomplex` to minimize the amount of modification? (this also keeps the original line numbers so it's better for cuda-gdb)

Reply:
Thank you very much for your understanding and detailed comments, @kmaehashi. I'm working on it. 

Two quick replies:
> Could you make the option a keyword-only argument so that we can easily consider reorder/renaming later?

If I understand you correctly, I don't think for `compile_with_cache` this is possible unfortunately. `cdef` and `cpdef` functions do not support `**kwargs` in the end of the function signature (will get an `Expected ')', found '**'` error).

> this also keeps the original line numbers so it's better for cuda-gdb

This is a very good suggestion which I would never come up myself! I never used `cuda-gdb` after all...

Reply:
> If I understand you correctly, I don't think for compile_with_cache this is possible unfortunately. cdef and cpdef functions do not support **kwargs in the end of the function signature (will get an Expected ')', found '**' error).

Yes, you're correct. I think it's OK for `compile_with_cache` as it is not public (documented) interface.